arangod/IResearch/IResearchAnalyzerFeature.cpp

```diff
@@ -458,22 +471,87 @@ arangodb::aql::AqlValue aqlFnTokens(arangodb::aql::ExpressionContext* expression
 
   arangodb::velocypack::Builder builder(*buffer);
-
   builder.openArray();
 
-  while (analyzer->next()) {
-    auto value = irs::ref_cast<char>(values->value());
-
-    arangodb::iresearch::addStringRef(builder, value);
-  }
+  std::deque<arangodb::velocypack::ArrayIterator> arrayIteratorStack;
+  auto current = args[0].slice();
+  do {
+    // stack opening non-empty arrays
+    while (current.isArray() && !current.isEmptyArray()) {
+      arrayIteratorStack.emplace_back(current);
+      builder.openArray();
+      current = arrayIteratorStack.back().value();
+    }
+    // process current item
+    if (current.isString()) {
+      auto const& data = arangodb::iresearch::getStringRef(current);
+      if (!string_analyzer->reset(data)) {
+        auto const message = "failure to reset arangosearch analyzer: ' "s +
+                             static_cast<std::string>(name) +
+                             "' while computing result for function 'TOKENS'";
+
+        LOG_TOPIC("45a2d", WARN, arangodb::iresearch::TOPIC) << message;
+        THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL, message);
+      }
+      while (string_analyzer->next()) {
+        auto value = irs::ref_cast<char>(string_terms->value());
+        arangodb::iresearch::addStringRef(builder, value);
+      }
+    } else if (current.isBool()) {
+      arangodb::iresearch::addStringRef(
+          builder, arangodb::basics::StringUtils::encodeBase64(irs::ref_cast<char>(
+                       irs::boolean_token_stream::value(current.getBoolean()))));
+    } else if (current.isNull()) {
+      arangodb::iresearch::addStringRef(
+          builder, arangodb::basics::StringUtils::encodeBase64(
+                       irs::ref_cast<char>(irs::null_token_stream::value_null())));
+    } else if (current.isNumber()) {
+      TRI_ASSERT(current.isDouble() || current.isInteger());
+      if (current.isDouble()) {
+        numeric_analyzer.reset(current.getDouble());
+      } else if (current.isInteger()) {
+        numeric_analyzer.reset(current.getInt());
+      }
+      while (numeric_analyzer.next()) {
+        arangodb::iresearch::addStringRef(builder,
+                                          arangodb::basics::StringUtils::encodeBase64(
+                                              irs::ref_cast<char>(numeric_terms->value())));
+      }
+    } else if (current.isEmptyArray()){ 
+      // empty array in = empty array out
+      builder.openArray();
+      builder.close();
+    } else {
+      auto const message = "unexpected parameter type '"s + current.typeName() +
+                           "' while computing result for function 'TOKENS'";
+      LOG_TOPIC("45a2e", WARN, arangodb::iresearch::TOPIC) << message;
+      THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_BAD_PARAMETER, message);
+    }
+    // de-stack all closing arrays
+    while (!arrayIteratorStack.empty()) {
+      auto& currentArrayIterator = arrayIteratorStack.back();
+      if (!currentArrayIterator.isLast()) {
```

Comment:
I think we can avoid checking `isLast(...)` before `next(...)`, the latter returns `false` if there are no more elements

Reply:
no, next is void and there will be exception on following value() call, or we could just check valid() which is not faster than just check isLast in advance